File: driver-sync/src/main/com/mongodb/client/ListSearchIndexesIterable.java
/*
 * Copyright 2008-present MongoDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mongodb.client;

import com.mongodb.ExplainVerbosity;
import com.mongodb.annotations.Alpha;
import com.mongodb.annotations.Evolving;
import com.mongodb.annotations.Reason;
import com.mongodb.client.cursor.TimeoutMode;
import com.mongodb.client.model.Collation;
import com.mongodb.lang.Nullable;
import org.bson.BsonValue;
import org.bson.Document;

import java.util.concurrent.TimeUnit;
/**
 * Iterable for listing Atlas Search indexes.
 * This interface contains aggregate options and that of applied to {@code $listSearchIndexes} operation.
 *
 * @param <TResult> The type of the result.
 * @mongodb.driver.manual reference/operator/aggregation/listSearchIndexes ListSearchIndexes
 * @since 4.11
 * @mongodb.server.release 6.0
 */
@Evolving
public interface ListSearchIndexesIterable<TResult> extends MongoIterable<TResult> {

    /**
     * Sets the index name for this operation.
     *
     * @param indexName the index name.
     * @return this.
     */
    ListSearchIndexesIterable<TResult> name(String indexName);

    /**
     * Enables writing to temporary files. A null value indicates that it's unspecified.
     *
     * @param allowDiskUse true if writing to temporary files is enabled.
     * @return this.
     * @mongodb.driver.manual reference/command/aggregate/ Aggregation
     */
    ListSearchIndexesIterable<TResult> allowDiskUse(@Nullable Boolean allowDiskUse);

    /**
     * Sets the number of documents to return per batch.
     *
     * @param batchSize the batch size.
     * @return this.
     * @mongodb.driver.manual reference/method/cursor.batchSize/#cursor.batchSize Batch Size
     */
    ListSearchIndexesIterable<TResult> batchSize(int batchSize);

    /**
     * Sets the maximum execution time on the server for this operation.
     *
     * @param maxTime  the max time.
     * @param timeUnit the time unit, which may not be null.
     * @return this.
     * @mongodb.driver.manual reference/method/cursor.maxTimeMS/#cursor.maxTimeMS Max Time
     */
    ListSearchIndexesIterable<TResult> maxTime(long maxTime, TimeUnit timeUnit);

    /**
     * Sets the collation options
     *
     * <p>A null value represents the server default.</p>
     *
     * @param collation the collation options to use
     * @return this
     */
    ListSearchIndexesIterable<TResult> collation(@Nullable Collation collation);

    /**
     * Sets the comment for this operation. A null value means no comment is set.
     *
     * @param comment the comment.
     * @return this
     */
    ListSearchIndexesIterable<TResult> comment(@Nullable String comment);

    /**
     * Sets the comment for this operation. A null value means no comment is set.
     *
     * @param comment the comment.
     * @return this.
     */
    ListSearchIndexesIterable<TResult> comment(@Nullable BsonValue comment);

    /**
     * Sets the timeoutMode for the cursor.
     *
     * <p>
     *     Requires the {@code timeout} to be set, either in the {@link com.mongodb.MongoClientSettings},
     *     via {@link MongoDatabase} or via {@link MongoCollection}
     * </p>
     * @param timeoutMode the timeout mode
     * @return this
     * @since 5.2
     */
    @Alpha(Reason.CLIENT)
    ListSearchIndexesIterable<TResult> timeoutMode(TimeoutMode timeoutMode);

    /**
     * Explain the execution plan for this operation with the server's default verbosity level.
     *
     * @return the execution plan.
     * @mongodb.driver.manual reference/command/explain/
     */
    Document explain();

    /**
     * Explain the execution plan for this operation with the given verbosity level.
     *
     * @param verbosity the verbosity of the explanation.
     * @return the execution plan.
     * @mongodb.driver.manual reference/command/explain/
     */
    Document explain(ExplainVerbosity verbosity);

    /**
     * Explain the execution plan for this operation with the server's default verbosity level.
     *
     * @param <E>                the type of the document class.
     * @param explainResultClass the document class to decode into.
     * @return the execution plan.
     * @mongodb.driver.manual reference/command/explain/
     */
    <E> E explain(Class<E> explainResultClass);

    /**
     * Explain the execution plan for this operation with the given verbosity level.
     *
     * @param <E>                the type of the document class.
     * @param explainResultClass the document class to decode into.
     * @param verbosity          the verbosity of the explanation.
     * @return the execution plan.
     * @mongodb.driver.manual reference/command/explain/
     */
    <E> E explain(Class<E> explainResultClass, ExplainVerbosity verbosity);
}

File: driver-kotlin-sync/src/main/kotlin/com/mongodb/kotlin/client/ListSearchIndexesIterable.kt
/*
 * Copyright 2008-present MongoDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mongodb.kotlin.client

import com.mongodb.ExplainVerbosity
import com.mongodb.client.ListSearchIndexesIterable as JListSearchIndexesIterable
import com.mongodb.client.cursor.TimeoutMode
import com.mongodb.client.model.Collation
import java.util.concurrent.TimeUnit
import org.bson.BsonValue
import org.bson.Document

/**
 * Iterable like implementation for list Atlas Search index operations.
 *
 * @param T The type of the result.
 * @see [List indexes](https://www.mongodb.com/docs/manual/reference/command/listIndexes/)
 */
public class ListSearchIndexesIterable<T : Any>(private val wrapped: JListSearchIndexesIterable<T>) :
    MongoIterable<T>(wrapped) {

    public override fun batchSize(batchSize: Int): ListSearchIndexesIterable<T> {
        super.batchSize(batchSize)
        return this
    }

    /**
     * Sets the timeoutMode for the cursor.
     *
     * Requires the `timeout` to be set, either in the [com.mongodb.MongoClientSettings], via [MongoDatabase] or via
     * [MongoCollection]
     *
     * @param timeoutMode the timeout mode
     * @return this
     * @since CSOT
     */
    public fun timeoutMode(timeoutMode: TimeoutMode): ListSearchIndexesIterable<T> {
        wrapped.timeoutMode(timeoutMode)
        return this
    }

    /**
     * Sets an Atlas Search index name for this operation.
     *
     * @param indexName Atlas Search index name.
     * @return this.
     */
    public fun name(indexName: String): ListSearchIndexesIterable<T> = apply { wrapped.name(indexName) }

    /**
     * Enables writing to temporary files. A null value indicates that it's unspecified.
     *
     * @param allowDiskUse true if writing to temporary files is enabled.
     * @return this.
     * @see [Aggregation command](https://www.mongodb.com/docs/manual/reference/command/aggregate/)
     */
    public fun allowDiskUse(allowDiskUse: Boolean?): ListSearchIndexesIterable<T> = apply {
        wrapped.allowDiskUse(allowDiskUse)
    }

    /**
     * Sets the maximum execution time on the server for this operation.
     *
     * **NOTE**: The maximum execution time option is deprecated. Prefer using the operation execution timeout
     * configuration options available at the following levels:
     * - [com.mongodb.MongoClientSettings.Builder.timeout]
     * - [MongoDatabase.withTimeout]
     * - [MongoCollection.withTimeout]
     * - [ClientSession]
     *
     * When executing an operation, any explicitly set timeout at these levels takes precedence, rendering this maximum
     * execution time irrelevant. If no timeout is specified at these levels, the maximum execution time will be used.
     *
     * @param maxTime the max time.
     * @param timeUnit the time unit, defaults to Milliseconds.
     * @return this.
     */
    @Deprecated("Prefer using the operation execution timeout configuration option", level = DeprecationLevel.WARNING)
    @Suppress("DEPRECATION")
    public fun maxTime(maxTime: Long, timeUnit: TimeUnit = TimeUnit.MILLISECONDS): ListSearchIndexesIterable<T> =
        apply {
            wrapped.maxTime(maxTime, timeUnit)
        }

    /**
     * Sets the collation options.
     *
     * A null value represents the server default.
     *
     * @param collation the collation options to use.
     * @return this.
     */
    public fun collation(collation: Collation?): ListSearchIndexesIterable<T> = apply { wrapped.collation(collation) }

    /**
     * Sets the comment for this operation. A null value means no comment is set.
     *
     * @param comment the comment.
     * @return this.
     */
    public fun comment(comment: String?): ListSearchIndexesIterable<T> = apply { wrapped.comment(comment) }

    /**
     * Sets the comment for this operation. A null value means no comment is set.
     *
     * @param comment the comment.
     * @return this.
     */
    public fun comment(comment: BsonValue?): ListSearchIndexesIterable<T> = apply { wrapped.comment(comment) }

    /**
     * Explain the execution plan for this operation with the given verbosity level.
     *
     * @param verbosity the verbosity of the explanation.
     * @return the execution plan.
     * @see [Explain command](https://www.mongodb.com/docs/manual/reference/command/explain/)
     */
    public fun explain(verbosity: ExplainVerbosity? = null): Document = explain<Document>(verbosity)

    /**
     * Explain the execution plan for this operation with the given verbosity level.
     *
     * @param R the type of the document class.
     * @param resultClass the result document type.
     * @param verbosity the verbosity of the explanation.
     * @return the execution plan.
     * @see [Explain command](https://www.mongodb.com/docs/manual/reference/command/explain/)
     */
    public fun <R : Any> explain(resultClass: Class<R>, verbosity: ExplainVerbosity? = null): R =
        if (verbosity == null) wrapped.explain(resultClass) else wrapped.explain(resultClass, verbosity)

    /**
     * Explain the execution plan for this operation with the given verbosity level.
     *
     * @param R the type of the document class.
     * @param verbosity the verbosity of the explanation.
     * @return the execution plan.
     * @see [Explain command](https://www.mongodb.com/docs/manual/reference/command/explain/)
     */
    public inline fun <reified R : Any> explain(verbosity: ExplainVerbosity? = null): R =
        explain(R::class.java, verbosity)
}
